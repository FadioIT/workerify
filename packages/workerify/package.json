{
  "name": "@fadioit/workerify",
  "version": "0.1.0",
  "main": "lib/index.js",
  "license": "MIT",
  "scripts": {
    "build": "[ -e lib ] && rm -rf lib; mkdir lib; babel src --out-dir lib"
  },
  "devDependencies": {
    "@babel/cli": "^7.1.2",
    "@babel/core": "^7.1.2",
    "@babel/preset-env": "^7.1.0"
  }
}
