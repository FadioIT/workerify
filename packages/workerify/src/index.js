export const createWorker = (source, isIterator) => {
  const deferred = createDeferred();
  const bridge = isIterator
    ? createIteratorBridge(deferred)
    : createBridge(deferred);

  const workerSource = isIterator
    ? `
  var worker = ${source};
  var iterator;
  self.onmessage = function (event) {
    if (event.data.method === 'init') {
      iterator = worker(event.data.value);
    } else {
      Promise.resolve(iterator[event.data.method](event.data.value))
        .then(postMessage, close)
    }
  }`
    : `
  var worker = ${source};
  self.onmessage = function (event) {
    Promise.resolve(worker(event.data))
      .then(postMessage, close)
  }`;

  if (window.Worker) {
    const worker = new Worker(
      URL.createObjectURL(new Blob([workerSource]), {
        type: 'application/javascript',
      }),
    );
    deferred.resolve(worker);
    bridge.terminate = () => worker.terminate();
  }

  return bridge;
};

export const createIteratorBridge = deferred => (value, transferable) => {
  const bridge = createBridge(deferred);
  bridge({ method: 'init', value }, transferable);

  return {
    next: (value, transferable) =>
      bridge({ method: 'next', value }, transferable),
    return: value => bridge({ method: 'return', value }),
    throw: value => bridge({ method: 'throw', value }),
  };
};

export const createBridge = deferred => async (value, transferable) => {
  const worker = await deferred.promise;
  const handlers = [];

  worker.onerror = event => {
    handlers.forEach(([, reject]) => reject(event.error));
  };
  worker.onmessage = event => {
    handlers.forEach(([resolve]) => resolve(event.data));
  };

  const result = new Promise((resolve, reject) => {
    handlers.push([resolve, reject]);
  });

  worker.postMessage(value, transferable);
  return result;
};

export const createDeferred = () => {
  const deferred = {};
  deferred.promise = new Promise((resolve, reject) => {
    deferred.resolve = resolve;
    deferred.reject = reject;
  });
  return deferred;
};

export default createWorker;
