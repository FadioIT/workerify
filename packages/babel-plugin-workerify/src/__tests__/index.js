import { transformSync } from '@babel/core';
// import babelPresetEnv from '@babel/preset-env';
import plugin, { WORKERIFY_MODULE_NAME } from '../';

const transform = source =>
  transformSync(source, { plugins: [plugin], presets: ['@babel/preset-env'] });

const arrowFunctionString = `(a, b) => {
  const [c] = require('some-module')(a, b);
  return c;
}`;
const functionString = `function (a, b) {
  const [c] = require('some-module')(a, b);
  return c;
}`;
const generatorString = `function*(a, b) {
  const [c] = require('some-module')(a, b);
  return c;
}`;

describe('stringifies callback in place', () => {
  it('stringify functions', () => {
    const source = `
      import workerify from '${WORKERIFY_MODULE_NAME}';

      workerify(${functionString});
    `;

    const { code, ast } = transform(source, { plugins: [plugin] });
    expect(code).toMatchSnapshot();
  });

  it('stringify arrow functions', () => {
    const source = `
      import workerify from '${WORKERIFY_MODULE_NAME}';

      workerify(${arrowFunctionString});
    `;

    const { code, ast } = transform(source, { plugins: [plugin] });
    expect(code).toMatchSnapshot();
  });

  it('stringify generators', () => {
    const source = `
      import workerify from '${WORKERIFY_MODULE_NAME}';

      workerify(${generatorString});
    `;

    const { code } = transform(source, { plugins: [plugin] });
    expect(code).toMatchSnapshot();
  });
});

describe('replaces references with their stringified value', () => {
  it('replace functions', () => {
    const source = `
      import workerify from '${WORKERIFY_MODULE_NAME}';

      const workerFunc = ${functionString};
      workerify(workerFunc);
    `;

    const { code } = transform(source, { plugins: [plugin] });
    expect(code).toMatchSnapshot();
  });

  it('replace arrow functions', () => {
    const source = `
      import workerify from '${WORKERIFY_MODULE_NAME}';

      const workerFunc = ${arrowFunctionString};
      workerify(workerFunc);
    `;

    const { code } = transform(source, { plugins: [plugin] });
    expect(code).toMatchSnapshot();
  });

  it('replace generators', () => {
    const source = `
      import workerify from '${WORKERIFY_MODULE_NAME}';

      const workerFunc = ${generatorString};
      workerify(workerFunc);
    `;

    const { code } = transform(source, { plugins: [plugin] });
    expect(code).toMatchSnapshot();
  });
});
