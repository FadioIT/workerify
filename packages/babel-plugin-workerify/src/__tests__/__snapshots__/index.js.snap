// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replaces references with their stringified value replace arrow functions 1`] = `
"\\"use strict\\";

var _workerify = _interopRequireDefault(require(\\"@fadioit/workerify\\"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError(\\"Invalid attempt to destructure non-iterable instance\\"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i[\\"return\\"] != null) _i[\\"return\\"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var workerFunc = function workerFunc(a, b) {
  var _require = require('some-module')(a, b),
      _require2 = _slicedToArray(_require, 1),
      c = _require2[0];

  return c;
};

(0, _workerify.default)(\\"\\\\\\"use strict\\\\\\";\\\\n\\\\nfunction _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }\\\\n\\\\nfunction _nonIterableRest() { throw new TypeError(\\\\\\"Invalid attempt to destructure non-iterable instance\\\\\\"); }\\\\n\\\\nfunction _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i[\\\\\\"return\\\\\\"] != null) _i[\\\\\\"return\\\\\\"](); } finally { if (_d) throw _e; } } return _arr; }\\\\n\\\\nfunction _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }\\\\n\\\\n(function (a, b) {\\\\n  var _require = require('some-module')(a, b),\\\\n      _require2 = _slicedToArray(_require, 1),\\\\n      c = _require2[0];\\\\n\\\\n  return c;\\\\n});\\", false);"
`;

exports[`replaces references with their stringified value replace functions 1`] = `
"\\"use strict\\";

var _workerify = _interopRequireDefault(require(\\"@fadioit/workerify\\"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError(\\"Invalid attempt to destructure non-iterable instance\\"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i[\\"return\\"] != null) _i[\\"return\\"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var workerFunc = function workerFunc(a, b) {
  var _require = require('some-module')(a, b),
      _require2 = _slicedToArray(_require, 1),
      c = _require2[0];

  return c;
};

(0, _workerify.default)(\\"\\\\\\"use strict\\\\\\";\\\\n\\\\nfunction _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }\\\\n\\\\nfunction _nonIterableRest() { throw new TypeError(\\\\\\"Invalid attempt to destructure non-iterable instance\\\\\\"); }\\\\n\\\\nfunction _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i[\\\\\\"return\\\\\\"] != null) _i[\\\\\\"return\\\\\\"](); } finally { if (_d) throw _e; } } return _arr; }\\\\n\\\\nfunction _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }\\\\n\\\\n(function (a, b) {\\\\n  var _require = require('some-module')(a, b),\\\\n      _require2 = _slicedToArray(_require, 1),\\\\n      c = _require2[0];\\\\n\\\\n  return c;\\\\n});\\", false);"
`;

exports[`replaces references with their stringified value replace generators 1`] = `
"\\"use strict\\";

var _workerify = _interopRequireDefault(require(\\"@fadioit/workerify\\"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError(\\"Invalid attempt to destructure non-iterable instance\\"); }

function _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i[\\"return\\"] != null) _i[\\"return\\"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var workerFunc =
/*#__PURE__*/
regeneratorRuntime.mark(function workerFunc(a, b) {
  var _require, _require2, c;

  return regeneratorRuntime.wrap(function workerFunc$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          _require = require('some-module')(a, b), _require2 = _slicedToArray(_require, 1), c = _require2[0];
          return _context.abrupt(\\"return\\", c);

        case 2:
        case \\"end\\":
          return _context.stop();
      }
    }
  }, workerFunc, this);
});
(0, _workerify.default)(\\"\\\\\\"use strict\\\\\\";\\\\n\\\\nfunction _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }\\\\n\\\\nfunction _nonIterableRest() { throw new TypeError(\\\\\\"Invalid attempt to destructure non-iterable instance\\\\\\"); }\\\\n\\\\nfunction _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i[\\\\\\"return\\\\\\"] != null) _i[\\\\\\"return\\\\\\"](); } finally { if (_d) throw _e; } } return _arr; }\\\\n\\\\nfunction _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }\\\\n\\\\n/*#__PURE__*/\\\\nregeneratorRuntime.mark(function _callee(a, b) {\\\\n  var _require, _require2, c;\\\\n\\\\n  return regeneratorRuntime.wrap(function _callee$(_context) {\\\\n    while (1) {\\\\n      switch (_context.prev = _context.next) {\\\\n        case 0:\\\\n          _require = require('some-module')(a, b), _require2 = _slicedToArray(_require, 1), c = _require2[0];\\\\n          return _context.abrupt(\\\\\\"return\\\\\\", c);\\\\n\\\\n        case 2:\\\\n        case \\\\\\"end\\\\\\":\\\\n          return _context.stop();\\\\n      }\\\\n    }\\\\n  }, _callee, this);\\\\n});\\", false);"
`;

exports[`stringifies callback in place stringify arrow functions 1`] = `
"\\"use strict\\";

var _workerify = _interopRequireDefault(require(\\"@fadioit/workerify\\"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

(0, _workerify.default)(\\"\\\\\\"use strict\\\\\\";\\\\n\\\\nfunction _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }\\\\n\\\\nfunction _nonIterableRest() { throw new TypeError(\\\\\\"Invalid attempt to destructure non-iterable instance\\\\\\"); }\\\\n\\\\nfunction _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i[\\\\\\"return\\\\\\"] != null) _i[\\\\\\"return\\\\\\"](); } finally { if (_d) throw _e; } } return _arr; }\\\\n\\\\nfunction _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }\\\\n\\\\n(function (a, b) {\\\\n  var _require = require('some-module')(a, b),\\\\n      _require2 = _slicedToArray(_require, 1),\\\\n      c = _require2[0];\\\\n\\\\n  return c;\\\\n});\\", false);"
`;

exports[`stringifies callback in place stringify functions 1`] = `
"\\"use strict\\";

var _workerify = _interopRequireDefault(require(\\"@fadioit/workerify\\"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

(0, _workerify.default)(\\"\\\\\\"use strict\\\\\\";\\\\n\\\\nfunction _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }\\\\n\\\\nfunction _nonIterableRest() { throw new TypeError(\\\\\\"Invalid attempt to destructure non-iterable instance\\\\\\"); }\\\\n\\\\nfunction _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i[\\\\\\"return\\\\\\"] != null) _i[\\\\\\"return\\\\\\"](); } finally { if (_d) throw _e; } } return _arr; }\\\\n\\\\nfunction _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }\\\\n\\\\n(function (a, b) {\\\\n  var _require = require('some-module')(a, b),\\\\n      _require2 = _slicedToArray(_require, 1),\\\\n      c = _require2[0];\\\\n\\\\n  return c;\\\\n});\\", false);"
`;

exports[`stringifies callback in place stringify generators 1`] = `
"\\"use strict\\";

var _workerify = _interopRequireDefault(require(\\"@fadioit/workerify\\"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

(0, _workerify.default)(\\"\\\\\\"use strict\\\\\\";\\\\n\\\\nfunction _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _nonIterableRest(); }\\\\n\\\\nfunction _nonIterableRest() { throw new TypeError(\\\\\\"Invalid attempt to destructure non-iterable instance\\\\\\"); }\\\\n\\\\nfunction _iterableToArrayLimit(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i[\\\\\\"return\\\\\\"] != null) _i[\\\\\\"return\\\\\\"](); } finally { if (_d) throw _e; } } return _arr; }\\\\n\\\\nfunction _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }\\\\n\\\\n/*#__PURE__*/\\\\nregeneratorRuntime.mark(function _callee(a, b) {\\\\n  var _require, _require2, c;\\\\n\\\\n  return regeneratorRuntime.wrap(function _callee$(_context) {\\\\n    while (1) {\\\\n      switch (_context.prev = _context.next) {\\\\n        case 0:\\\\n          _require = require('some-module')(a, b), _require2 = _slicedToArray(_require, 1), c = _require2[0];\\\\n          return _context.abrupt(\\\\\\"return\\\\\\", c);\\\\n\\\\n        case 2:\\\\n        case \\\\\\"end\\\\\\":\\\\n          return _context.stop();\\\\n      }\\\\n    }\\\\n  }, _callee, this);\\\\n});\\", true);"
`;
